lib/pulse_core/PulseCore.Atomic.fst

```diff
@@ -0,0 +1,313 @@
+  : squash (pure p == emp)
+  = calc (==) {
+      pure p;
+    (==) { equiv (pure_equiv p True ()) }
+      pure True;
+    (==) { equiv (A.pure_true ()) }
+      emp;
+  }
+
+let emp_unit_r (p:slprop)
+: squash (p ** emp == p)
+= calc (==) {
+    (p ** emp);
+   (==) { equiv (slprop_equiv_comm p emp) }
+    (emp ** p);
+   (==) { equiv (slprop_equiv_unit p) }
+     p;
+  }
+
+let return_atomic' #a x post
+: stt_atomic a #Unobservable emp_inames
+      (post x ** pure (x == x))
+      (fun r -> post r ** pure (r == x))
+= A.return #a #(fun r -> post r ** pure (r == x)) x
+
+let return_atomic #a x post
+: stt_atomic a #Unobservable emp_inames
+      (post x)
+      (fun r -> post r ** pure (r == x))
+= emp_unit_r (post x);
+  pure_trivial (x == x) ();
+  coerce_eq () (return_atomic' #a x post)
+
+let return_atomic_noeq #a x post = A.return #a #post x
+
+let bind_atomic
+    (#a:Type u#a)
+    (#b:Type u#b)
+    (#obs1:_)
+    (#obs2:observability { at_most_one_observable obs1 obs2 })
+    (#opens:inames)
+    (#pre1:slprop)
+    (#post1:a -> slprop)
+    (#post2:b -> slprop)
+    (e1:stt_atomic a #obs1 opens pre1 post1)
+    (e2:(x:a -> stt_atomic b #obs2 opens (post1 x) post2))
+= A.bind e1 e2
+
+let lift_unobservable
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+= e
+
+let lift_atomic0
+    (#a:Type u#0)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift0 e
+
+let lift_atomic1
+    (#a:Type u#1)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift1 e
+
+let lift_atomic2
+    (#a:Type u#2)
+    (#obs:_)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt a pre post
+= A.lift2 e
+
+let frame_atomic
+    (#a:Type u#a)
+    (#obs: observability)
+    (#opens:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (frame:slprop)
+    (e:stt_atomic a #obs opens pre post)
+: stt_atomic a #obs opens (pre ** frame) (fun x -> post x ** frame)
+= A.frame e
+
+let sub_atomic
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens:inames)
+    (#pre1:slprop)
+    (pre2:slprop)
+    (#post1:a -> slprop)
+    (post2:a -> slprop)
+    (pf1 : slprop_equiv pre1 pre2)
+    (pf2 : slprop_post_equiv post1 post2)
+    (e:stt_atomic a #obs opens pre1 post1)
+: stt_atomic a #obs opens pre2 post2
+= A.sub pre2 post2 e
+
+let sub_invs_stt_atomic
+    (#a:Type u#a)
+    (#obs:_)
+    (#opens1 #opens2:inames)
+    (#pre:slprop)
+    (#post:a -> slprop)
+    (e:stt_atomic a #obs opens1 pre post)
+    (_ : squash (inames_subset opens1 opens2))
+: stt_atomic a #obs opens2 pre post
+= assert (Set.equal (Set.union opens1 opens2) opens2);
+  A.weaken opens2 e
+
+let stt_ghost a opens pre post = Ghost.erased (act a opens pre post)

```

Comment:
An stt_ghost is just an erased action, which is simlar to the representation used in Steel.

I am thinking to refine this in the future to actions that have no write effects on concrete memory locations. Forbidding read effects on concrete memory locations would be nice too, though that would require a relational spec, which is harder, since the underlying mst monad is not relational.